usr.bin/mkstr/mkstr.1
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)mkstr.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: head/usr.bin/mkstr/mkstr.1 284060 2015-06-06 12:08:10Z bapt $
.\"
.Dd June 6, 2015
.Dt MKSTR 1
.Os
.Sh NAME
.Nm mkstr
.Nd create an error message file by massaging C source
.Sh SYNOPSIS
.Nm
.Op Fl
.Ar mesgfile
.Ar prefix Ar
.Sh DESCRIPTION
The
.Nm
utility creates a file containing error messages extracted from C source,
and restructures the same C source, to utilize the created error message
file.
The intent of
.Nm
was to reduce the size of large programs and
reduce swapping (see
.Sx BUGS
section below).
.Pp
The
.Nm
utility processes each of the specified files,
placing a restructured version of the input in a file whose name
consists of the specified
.Ar prefix
and the original name.
A typical usage of
.Nm
is
.Pp
.Dl "mkstr pistrings xx *.c"
.Pp
This command causes all the error messages from the C source
files in the current directory to be placed in the file
.Pa pistrings
and restructured copies of the sources to be placed in
files whose names are prefixed with
.Dq Li xx .
.Pp
Options:
.Bl -tag -width indent
.It Fl
Error messages are placed at the end of the specified
message file for recompiling part of a large
.Nm Ns ed
program.
.El
.Pp
The
.Nm
utility finds error messages in the source by
searching for the string
.Sq Li error("
in the input stream.
Each time it occurs, the C string starting at the
.Ql \&"
is stored
in the message file followed by a null character and a new-line character;
The new source is restructured with
.Xr lseek 2
pointers into the error message file for retrieval.
.Bd -literal -offset indent
char efilname = "/usr/lib/pi_strings";
int efil = -1;

error(a1, a2, a3, a4)
{
	char buf[256];

	if (efil < 0) {
		efil = open(efilname, 0);
		if (efil < 0)
			err(1, "%s", efilname);
	}
	if (lseek(efil, (off_t)a1, SEEK_SET) < 0 ||
	    read(efil, buf, 256) <= 0)
		err(1, "%s", efilname);
	printf(buf, a2, a3, a4);
}
.Ed
.Sh SEE ALSO
.Xr gencat 1 ,
.Xr xstr 1 ,
.Xr lseek 2
.Sh HISTORY
The
.Nm
utility first appeared in
.Bx 1 .
.Sh AUTHORS
.An -nosplit
.An Bill Joy
and
.An Chuck Haley ,
1977.
.Sh BUGS
The
.Nm
utility was intended for the limited architecture of the PDP 11 family.
Very few programs actually use it.
The memory savings are negligible in modern computers.


usr.bin/mkstr/mkstr.c
/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)mkstr.c	8.1 (Berkeley) 6/6/93";
#endif
#endif /* not lint */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: head/usr.bin/mkstr/mkstr.c 284063 2015-06-06 12:39:00Z bapt $");

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define	ungetchar(c)	ungetc(c, stdin)

/*
 * mkstr - create a string error message file by massaging C source
 *
 * Bill Joy UCB August 1977
 *
 * Modified March 1978 to hash old messages to be able to recompile
 * without addding messages to the message file (usually)
 *
 * Based on an earlier program conceived by Bill Joy and Chuck Haley
 *
 * Program to create a string error message file
 * from a group of C programs.  Arguments are the name
 * of the file where the strings are to be placed, the
 * prefix of the new files where the processed source text
 * is to be placed, and the files to be processed.
 *
 * The program looks for 'error("' in the source stream.
 * Whenever it finds this, the following characters from the '"'
 * to a '"' are replaced by 'seekpt' where seekpt is a
 * pointer into the error message file.
 * If the '(' is not immediately followed by a '"' no change occurs.
 *
 * The optional '-' causes strings to be added at the end of the
 * existing error message file for recompilation of single routines.
 */

static FILE	*mesgread, *mesgwrite;
static char	name[100], *np;

void copystr(void);
int fgetNUL(char *, int, FILE *);
unsigned hashit(char *, int, unsigned);
void inithash(void);
int match(const char *);
int octdigit(char);
void process(void);
void usage(void);

int
main(int argc, char *argv[])
{
	char addon = 0;
	size_t namelen;

	argc--, argv++;
	if (argc > 1 && argv[0][0] == '-')
		addon++, argc--, argv++;
	if (argc < 3)
		usage();
	mesgwrite = fopen(argv[0], addon ? "a" : "w");
	if (mesgwrite == NULL)
		err(1, "%s", argv[0]);
	mesgread = fopen(argv[0], "r");
	if (mesgread == NULL)
		err(1, "%s", argv[0]);
	inithash();
	argc--, argv++;
	namelen = strlcpy(name, argv[0], sizeof(name));
	if (namelen >= sizeof(name)) {
		errno = ENAMETOOLONG;
		err(1, "%s", argv[0]);
	}
	np = name + namelen;
	argc--, argv++;
	do {
		if (strlcpy(np, argv[0], sizeof(name) - namelen) >=
		    sizeof(name) - namelen) {
			errno = ENAMETOOLONG;
			err(1, "%s%s", name, argv[0]);
		}
		if (freopen(name, "w", stdout) == NULL)
			err(1, "%s", name);
		if (freopen(argv[0], "r", stdin) == NULL)
			err(1, "%s", argv[0]);
		process();
		argc--, argv++;
	} while (argc > 0);
	exit(0);
}

void
usage(void)
{
	fprintf(stderr, "usage: mkstr [-] mesgfile prefix file ...\n");
	exit(1);
}

void
process(void)
{
	int c;

	for (;;) {
		c = getchar();
		if (c == EOF)
			return;
		if (c != 'e') {
			putchar(c);
			continue;
		}
		if (match("error(")) {
			printf("error(");
			c = getchar();
			if (c != '"')
				putchar(c);
			else
				copystr();
		}
	}
}

int
match(const char *ocp)
{
	const char *cp;
	int c;

	for (cp = ocp + 1; *cp; cp++) {
		c = getchar();
		if (c != *cp) {
			while (ocp < cp)
				putchar(*ocp++);
			ungetchar(c);
			return (0);
		}
	}
	return (1);
}

void
copystr(void)
{
	int c, ch;
	char buf[512];
	char *cp = buf;

	for (;;) {
		if (cp == buf + sizeof(buf) - 2)
			errx(1, "message too long");
		c = getchar();
		if (c == EOF)
			break;
		switch (c) {

		case '"':
			*cp++ = 0;
			goto out;
		case '\\':
			c = getchar();
			switch (c) {

			case 'b':
				c = '\b';
				break;
			case 't':
				c = '\t';
				break;
			case 'r':
				c = '\r';
				break;
			case 'n':
				c = '\n';
				break;
			case '\n':
				continue;
			case 'f':
				c = '\f';
				break;
			case '0':
				c = 0;
				break;
			case '\\':
				break;
			default:
				if (!octdigit(c))
					break;
				c -= '0';
				ch = getchar();
				if (!octdigit(ch))
					break;
				c <<= 7, c += ch - '0';
				ch = getchar();
				if (!octdigit(ch))
					break;
				c <<= 3, c+= ch - '0', ch = -1;
				break;
			}
		}
		*cp++ = c;
	}
out:
	*cp = 0;
	printf("%d", hashit(buf, 1, 0));
}

int
octdigit(char c)
{

	return (c >= '0' && c <= '7');
}

void
inithash(void)
{
	char buf[512];
	int mesgpt = 0;

	rewind(mesgread);
	while (fgetNUL(buf, sizeof buf, mesgread) != 0) {
		hashit(buf, 0, mesgpt);
		mesgpt += strlen(buf) + 2;
	}
}

#define	NBUCKETS	511

static struct	hash {
	long	hval;
	unsigned hpt;
	struct	hash *hnext;
} *bucket[NBUCKETS];

unsigned
hashit(char *str, int really, unsigned fakept)
{
	int i;
	struct hash *hp;
	char buf[512];
	long hashval = 0;
	char *cp;

	if (really)
		fflush(mesgwrite);
	for (cp = str; *cp;)
		hashval = (hashval << 1) + *cp++;
	i = hashval % NBUCKETS;
	if (i < 0)
		i += NBUCKETS;
	if (really != 0)
		for (hp = bucket[i]; hp != 0; hp = hp->hnext)
		if (hp->hval == hashval) {
			fseek(mesgread, (long) hp->hpt, 0);
			fgetNUL(buf, sizeof buf, mesgread);
/*
			fprintf(stderr, "Got (from %d) %s\n", hp->hpt, buf);
*/
			if (strcmp(buf, str) == 0)
				break;
		}
	if (!really || hp == 0) {
		hp = (struct hash *) calloc(1, sizeof *hp);
		if (hp == NULL)
			err(1, NULL);
		hp->hnext = bucket[i];
		hp->hval = hashval;
		hp->hpt = really ? ftell(mesgwrite) : fakept;
		if (really) {
			fwrite(str, sizeof (char), strlen(str) + 1, mesgwrite);
			fwrite("\n", sizeof (char), 1, mesgwrite);
		}
		bucket[i] = hp;
	}
/*
	fprintf(stderr, "%s hashed to %ld at %d\n", str, hp->hval, hp->hpt);
*/
	return (hp->hpt);
}

int
fgetNUL(char *obuf, int rmdr, FILE *file)
{
	int c;
	char *buf = obuf;

	while (--rmdr > 0 && (c = getc(file)) != 0 && c != EOF)
		*buf++ = c;
	*buf++ = 0;
	getc(file);
	return ((feof(file) || ferror(file)) ? 0 : 1);
}
